### PR TITLE
LUCENE-10293: adds a stand-alone prepareSonatypeBundle that compiles …

### PR DESCRIPTION
**Pull request by dweiss**
#11329

### CODE REVIEW

File: lucene/distribution/build.gradle [hunk 1]
```diff
@@ -126,3 +126,31 @@ tasks.matching {it.name in [
...
+
+  destinationDirectory = file("${buildDir}/sonatype")
+  archiveFileName = "artifact-bundle-lucene.zip"
+
+  from({ ->
+    def releaseDir = propertyOrDefault("release.dir", null)
+    if (releaseDir == null) {
+      throw new GradleException("Pass an explicit release directory path with -Prelease.dir=...")
+    }
+    if (!file("${releaseDir}/.gitrev").exists()) {
```
Comment (jpountz) [inline lucene/distribution/build.gradle:142]: This file exists for me in `lucene/distribution/build/release` but it doesn't seem to exist in the release candidates that are produced by `dev-tools/scripts/buildAndPushRelease.py`.

Comment (dweiss) [reply lucene/distribution/build.gradle:142]: It should definitely be there?
https://github.com/apache/lucene/blob/branch_9_0/lucene/distribution/build.gradle#L79-L90

Comment (dweiss) [reply lucene/distribution/build.gradle]: I've modified the code to just check for 'maven' subfolder instead of the gitref.

Comment (dweiss) [reply lucene/distribution/build.gradle]: Just out of curiosity - did you use this task or ant, @jpountz ? Should I polish this and commit it in or do we want to work out some other solution?

**Comment by HoustonPutman** [comment]
Hey @dweiss I am finishing up the Solr 9.0.0 release and ran into this issue.

It looks like both Lucene 9.0 and 9.1 used the ant command from 8.11 to do the maven upload.

I tried using this method but unfortunately when you do the "upload bundle" there's a bug where it will add additional folder levels to the artifacts that you are uploading (if you have multiple artifacts included). I tried all possible ways I could think of around this, but nothing resulted in a correct repository.

After lots of looking, I found out that Spark has completely automated this part of their build: https://github.com/apache/spark/blob/v3.2.1/dev/create-release/release-build.sh

So I am building a script and releaseWizard entry to do this for Solr. Will happily port it to lucene afterwards.

Quick question though, do you think we should create the staging repository as a part of the release candidate, then delete it/promote it if the vote succeeds? If we did that, then we wouldn't have to add the maven artifacts to the release-candidate dist folder, and users could access the maven artifacts in a more natural way. I haven't used the staging repositories myself before, so I'm not sure if this is a good idea or not.

**Comment by dweiss** [comment]
> I tried using this method but unfortunately when you do the "upload bundle" there's a bug where it will add additional folder levels to the artifacts that you are uploading (if you have multiple artifacts included). I tried all possible ways I could think of around this, but nothing resulted in a correct repository.

I mentioned this somewhere, actually (and I provided a patch that did prepare the correct bundle structure). I'm not sure where this was and I have limited internet connectivity so I can't look it up... but I think it's here:

https://issues.apache.org/jira/browse/LUCENE-9488

Anyway. We already publish maven artifacts to nexus (snapshots). And the code to publish to releases is already there too -- see maven-to-nexus-releases.gradle and "mavenToApacheReleases" task. Why not just use this instead of some odd sh scripts?

**Comment by HoustonPutman** [comment]
I'm fairly confused. The `mavenToApacheReleases` task was authored in October of last year, while this PR is from December, so I assumed that there was an effort to solve this beyond using `mavenToApacheReleases`. Is this PR trying to tackle a different task than I think it is?

**Comment by dweiss** [comment]
Sorry. Working from remote/ holidays and only having 5 minutes to look at things. So the sonatype bundle should be working fine, really. I know about the problem with subfolders but I did have it working in other projects and I think it should be working here as well? Spark's solution seems terribly long and complicated to me (not to mention it's not really playing nice with Windows users, for example). I'd rather have it assemble a (portable) bundle in a portable way.

The problem, as far as I remember, was that the release process requires the same artifacts to be voted upon and then published. When you prepare a bundle, everything has to be flattened into a single level - this is a source of pain when reviewing artifacts for publication. And that's why you can't really use gradle's built-in mechanism of uploading artifacts (because they would/could be different on each gradle build from scratch). If we can live with a flatttened bundle though, it should be fairly easy to implement the release process around it.

**Comment by dweiss** [comment]
When you tried uploading the bundle, did you use this patch, @HoustonPutman ? Because it does "flatten" the artifact structure - 
```
  eachFile { FileCopyDetails details ->
      details.path = details.name
    }
```

and I recall I tried it with Apache Nexus and it seemed to work just fine.

**Comment by HoustonPutman** [comment]
I did use this patch, and it was all flattened. However Apache Nexus responded with a very long error message saying that every file in the zip was "invalid". (I then tried every possible combination of directories that I could try since flattened didn't work)

In the meantime I have a small script that I might use for 9.0.0. Agreed that I would like it to be portable, but for now this is the only way I've been able to get it to work. We can always move back to this when we find out how to get it to work.

It's much more digestible than the Spark script. You can check it out here: https://github.com/apache/solr/pull/807

**Comment by dweiss** [comment]
It is strange because I remember verifying that it worked... I'll take another look at this once I'm back from Easter break.

**Comment by dweiss** [comment]
I finally managed to take a look because this intrigued me (I knew this used to work). Turns out everything is ok with the patch - the bundle structure is fine when I build it with:
```
gradlew prepareSonatypeBundle -Pversion.release=10.0.0 -Psign=true
```
I can confirm Apache Nexus fails to accept it but it's because the ZIP file contains multiple artifacts (with multiple coordinates). If I limit the content of the ZIP to just one module (JAR+sources+javadocs+signatures), Apache Nexus accepts it and all the validation checks pass just fine. I didn't know about this limitation (one coordinate per bundle), but I can see it spelled here:

https://help.sonatype.com/repomanager2/staging-releases/artifact-bundles

In other words - these bundles are not going to be useful for us. :(
